packages/astro/src/cli/add.ts

```diff
@@ -0,0 +1,361 @@
+import { parseNpmName } from '../core/util.js';
+import { t, parse, visit, ensureImport, wrapDefaultExport, generate } from '../transform/index.js';
+
+export interface AddOptions {
+	logging: LogOptions;
+	cwd?: string;
+	flags: yargs.Arguments;
+}
+
+export interface IntegrationInfo {
+	id: string;
+	packageName: string;
+	dependencies: [name: string, version: string][];
+}
+
+const DEFAULT_CONFIG_STUB = `import { defineConfig } from 'astro/config';\n\nexport default defineConfig({});`;
+
+export async function add(names: string[], { cwd, flags, logging }: AddOptions) {
+	if (names.length === 0) {
+		error(logging, null, `\n${red('No integration specified!')}\n${dim('Try using')} astro add ${cyan('[name]')}`);
+		return;
+	}
+	const root = cwd ? path.resolve(cwd) : process.cwd();
+	let configURL = await resolveConfigURL({ cwd, flags });
+	applyPolyfill();
+	if (configURL) {
+		debug('add', `Found config at ${configURL}`);
+	} else {
+		info(logging, 'add', `Unable to locate a config file, generating one for you.`);
+		configURL = new URL('./astro.config.mjs', `file://${root}/`);
+		await fs.writeFile(fileURLToPath(configURL), DEFAULT_CONFIG_STUB, { encoding: 'utf-8' });
+	}
+
+	const integrations = await validateIntegrations(names);
+
+	// Add integrations to astro config
+	// TODO: At the moment, nearly nothing throws an error. We need more errors!
+	let ast: t.File | null = null;
+	try {
+		ast = await parseAstroConfig(configURL);
+
+		debug('add', 'Parsed astro config');
+
+		const defineConfig = t.identifier('defineConfig');
+		ensureImport(ast, t.importDeclaration([t.importSpecifier(defineConfig, defineConfig)], t.stringLiteral('astro/config')));
+		wrapDefaultExport(ast, defineConfig);
+
+		debug('add', 'Astro config ensured `defineConfig`');
+
+		for (const integration of integrations) {
+			await addIntegration(ast, integration);
+			debug('add', `Astro config added integration ${integration.id}`);
+		}
+	} catch (err) {
+		debug('add', 'Error parsing/modifying astro config: ', err);
+		info(
+			logging,
+			null,
+			"Sorry, we couldn't update your configuration automatically. [INSERT HOW TO DO IT MANUALLY --- this link might help: https://next--astro-docs-2.netlify.app/en/guides/integrations-guide/]"
+		);
+	}
+
+	let configResult: UpdateResult | undefined;
+	let installResult: UpdateResult | undefined;
+
+	if (ast) {
+		try {
+			configResult = await updateAstroConfig({ configURL, ast, logging });
+		} catch (err) {
+			debug('add', 'Error updating astro config', err);
+			error(logging, null, 'There has been an error updating the astro config. You might need to update it manually.');
+			return;
+		}
+	}
+
+	switch (configResult) {
+		case UpdateResult.cancelled: {
+			info(logging, null, msg.cancelled(`Your configuration has not been updated.`));
+			return;
+		}
+		case UpdateResult.none: {
+			info(logging, null, msg.success(`Configuration up-to-date. No changes needed!`));
```

Comment:
super nit: `pnpm` `npm` and `yarn` all seem happy with the shorter "Configuration up-to-date." style message, without a "No changes needed!" at the end. Maybe okay to cut that?